lib/cretonne/src/ssa.rs

```diff
@@ -0,0 +1,145 @@
+//! A SSA-building API that handles incomplete CFGs.
+
+/// The algorithm is based upon Braun M., Buchwald S., Hack S., Leißa R., Mallon C.,
+/// Zwinkau A. (2013) Simple and Efficient Construction of Static Single Assignment Form.
+/// In: Jhala R., De Bosschere K. (eds) Compiler Construction. CC 2013.
+/// Lecture Notes in Computer Science, vol 7791. Springer, Berlin, Heidelberg
+
+use ir::{Ebb, Function, Value, Type, ValueListPool, ValueList};
+use entity_map::{EntityMap, EntityRef};
+use flowgraph::ControlFlowGraph;
+use entity_list::ListPool;
+use sparse_map::{SparseMap, SparseMapValue};
+use packed_option::{PackedOption, ReservedValue};
+use std::u32;
+
+/// Structure containing the data relevant the construction of SSA for a given function.
+pub struct SSABuilder {
+    // Records for every variable its type and, for every revelant block, the last definition of
+    // the variable in the block.
+    variables: EntityMap<Variable, VariableData>,
+    // Records the position of the basic blocks and the list of values used but not defined in the
+    // block.
+    blocks: EntityMap<Block, BlockData>,
+    // Records the basic blocks at the beginning of the `Ebb`s.
+    ebb_headers: SparseMap<Ebb, (Ebb, Block)>,
```

Comment:
Indeed, that's the reason. I think this is the tradeoff between `EntityMap` and `SparseMap`, `SparseMap` takes less space if the keys are sparse but the values have to be larger to be able to compute the keys from them.